Make Sertra detail view use actual mod data

Oksamies opened the pull request:
refs TS-712

MythicManiac commented on apps/sertra/pages/server/[id].tsx:
```diff
@@ -30,8 +30,34 @@ export const getStaticProps: GetStaticProps<
+  const updated_listing_mods_data = [];
+  for (const mod_ref of data.mods) {
+    for (const mod of mods_data) {
+      const [mod_ref_owner, mod_ref_name, mod_ref_version] = mod_ref.split(
```
This isn't a valid way to parse package references unfortunately (I wish it was). The data includes a `full_name` property for each package version, so I'd suggest you
1. Check if the package `full_name` matches the start of the mod definition.
2. If yes, compare against all versions of the partially matching package until you find a full match.

MythicManiac replied:
If you really want to implement a parser, you can copy the implementation from here: https://github.com/thunderstore-io/Thunderstore/blob/master/django/thunderstore/repository/package_reference.py but I'd recommend against it as it's unnecessary complexity in this case.

Oksamies replied:
I'd like to do 1, but the problem is that owner and the mod name are flipped. I'll figure something out that doesn't need the parser and doesn't look like hell

Oksamies replied:
Hmm, hold on they aren't flipped, dunno what I thought yesterday, nvm.

Oksamies replied:
@MythicManiac Please go over the newest implementation. I did both ways, so that you(or someone) could tell me if it's a correctish way to do the parsing?

MythicManiac replied:
I'm still not sure why are you attempting to parse it instead of doing simple string comparison? If you do want to parse it this is definitely not how we want it written in code; there should be a standalone function/class handling the parsing rather than inlining it like it has been now written.

That said unless there's a reason you need to do the parsing, I'd just stick with the string comparison approach.

MythicManiac replied:
You should also move the package data fetching to some standalone class/function rather than inlining it to that specific view, chances are it's going to be re-used elsewhere too and it makes sense to cache it, which isn't really possible if it's inlined the way it is currently.

That's perhaps outside the scope of this PR however, but noting it anyway for future reference.

MythicManiac reviewed (changes requested):
See the comment thread here: https://github.com/thunderstore-io/thunderstore-ui/pull/516#discussion_r965154711

anttimaki commented on apps/sertra/pages/server/[id].tsx:
```diff
@@ -30,8 +30,34 @@ export const getStaticProps: GetStaticProps<
+      }
+    }
+  }
+  data.mods = updated_listing_mods_data;
```
I think mutating the `data`, which is a fetch response(?) is a bad idea. The processed data should be stored in state separately.

Oksamies replied:
Good point, dunno why I didn't just plug the new array into the return :D Fixd, this 👍

anttimaki commented on apps/sertra/pages/server/[id].tsx:
```diff
@@ -66,16 +92,15 @@ const ServerDetail: React.FC<{ detail_listing: ServerListingDetailData }> = ({
                   key={x.name}
                   name={x.name}
-                  description={x.description}
+                  owner={x.owner}
```
When object keys nicely match component props, you can do something like this:
`<ModCard key={x.name} {...x} description={x.description ?? null} />`

`{...x}` will override any props defined before it with values from `x`. Since there's no `x.key`, the key prop won't be overwritten.
Props defined after `{...x}` will override any props defined by `{...x}`. So even though there's `x.description`, redefining the prop will overwrite the value received from `x`. **Note**: I can't recall if this works here since the type of `x.description` might not match the type of `ModCard.description`. Worth checking out anyway, or at least for you to know in the future (in case you did not know this already).

Oksamies replied:
Seems like `<ModCard key={x.name} {...x} />` does work here, just had to define type for `x` correctly. Didn't know this way was possible. Definitely is a lot simpler and cleaner representation.

anttimaki replied:
Yeah, it's simpler, but use your own judgement when to use it. Since it obfuscates the props that are actually passed (or left out), it might be a bit confusing in some situations.

Also probably `props` or `modProps` would be a better variable name than `x` now that there's not that much repetition, but that's just an opinion.

Mark this comment resolved once you've read it.

Oksamies replied:
Changed the name 👍 And good point, I'll have to keep it mind, when using some more complex component.

anttimaki commented on apps/sertra/components/ModCard.tsx:
```diff
@@ -17,7 +17,11 @@ export const ModCard: React.FC<ModCard> = ({
+      {icon_url ? (
+        <img className={styles.imageIcon} src={icon_url}></img>
+      ) : (
+        <img className={styles.imageIcon} src="/favicon.ico"></img>
```
Isn't that `favicon.ico` the Vercel's default icon? Or has that been replaced already. Anyhow I'd prefer `public/ts-logo.svg` as a quick solution. Proper solution is probably asking grav whether we want to use the same approach as the new upcoming UI, or have some other placeholder for sertra.

Also since the only thing changing seems to be the `src`, the ternary operator is unnecessarily verbose. This would be more readable IMO:
`<img className={styles.imageIcon} src={icon_url ?? "/favicon.ico"} />`

Also also I think you should fix the things ESLint is complaining about?

Oksamies replied:
Ah I didn't know I could do that one line ternary thing 👍 Also switched to using the nexts Image component. I'm not sure what are the benefits of that, but at least they had an example on nextjs.com.

anttimaki commented on apps/sertra/components/ModCard.module.css:
```diff
@@ -17,7 +17,7 @@
 }
 
-.imagePlaceholder {
+.imageIcon {
```
"imageIcon" sounds stupid IMO. Maybe use "modImage" or "modIcon"?

anttimaki reviewed (commented):
My two cents. Didn't review the "parsing stuff" since you and Mythic seem to have that part under control/construction.

Oksamies commented:
I've addressed all the comments, but I'm hitting a wall with one of the suggestions Antti made (https://github.com/thunderstore-io/thunderstore-ui/pull/516#discussion_r965575108) please help, I don't have the mental energy to look at this anymore.

![image](https://user-images.githubusercontent.com/34424527/189214743-de0bd9d9-a32a-4708-a6dc-87aa9d9117ec.png)

MythicManiac commented on apps/sertra/components/PackageDataFetcher.tsx:
```diff
@@ -0,0 +1,65 @@
+        new_mod_data.owner = mod.owner;
+        new_mod_data.version = null;
+        for (const version of mod.versions) {
+          if (mod_ref.endsWith(version.version_number)) {
```
You **need** to do a full string comparison before you can be sure the mod actually matches, the current checks can let false positives through.

Likewise you can't assume the mod to match if just the start matches (on line 45), you can simply use that to reduce the amount of versions you enumerate over (just enumerate the mods with a matching start rather than _all_ versions). In other words you can't populate any data to the mod data before you have a full match, or otherwise you risk displaying invalid data.

I'd suggest you first flatten the entire mods list to a dictionary (you can use a [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) or a [record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type)), and then simply look up if a version exists there, rather than looping through everything multiple times.

MythicManiac commented on apps/sertra/components/PackageDataFetcher.tsx:
```diff
@@ -0,0 +1,65 @@
+        for (const version of mod.versions) {
+          if (mod_ref.endsWith(version.version_number)) {
+            new_mod_data.version = version.version_number;
+            new_mod_data.icon_url = version.icon ?? null;
```
At this point there should be no need to use the null coalescing operator, all versions have an icon and a description.

MythicManiac commented on apps/sertra/components/PackageDataFetcher.tsx:
```diff
@@ -0,0 +1,65 @@
+import { useEffect, useState } from "react";
```
This whole file is in the wrong location, and uses the wrong file extension.

The `components` directory should only be used for react components, and since this is simply typescript code, it should be placed elsewhere. Likewise there's no reason to use the `.tsx` extension (which is for react templates), use `.ts` instead (plain typescript).

What this actually is, is a hook, and it's fairly common to have one or more modules called `hooks` somewhere in a react project, which contain custom hook implementations.

MythicManiac commented on apps/sertra/components/ModCard.tsx:
```diff
@@ -2,21 +2,29 @@ import styles from "./ModCard.module.css";
+  owner: string;
   description: string;
+  version: string;
+  icon_url: string;
```
Assuming we get any mod data, this can never be null. All packages are required to have an icon.

MythicManiac commented on apps/sertra/components/PackageDataFetcher.tsx:
```diff
@@ -0,0 +1,65 @@
+
+export const FetchListingData = async (listingId: string) => {
+  const [listing, setListing] = useState<TempModRefType>();
+  useEffect(() => {
+    async function fetchListing() {
+      return await fetch(ApiURLs.ServerDetail(listingId));
+    }
+    async function jsonifyListing() {
+      setListing(await (await fetchListing()).json());
+    }
+    jsonifyListing();
+  });
```
Missing the second parameter (to the `useEffect` function), which should be `[listingId]` in this case. Look up how it works in case you're unfamiliar with what it does.

MythicManiac replied:
Why so verbose? There's no need to define named functions like this

```ts
useEffect(() => {
  fetch(ApiURLs.ServerDetail(listingId)).then((resp) => {
    setListing(resp.json());
  });
}, [listingId]);
```

Same applies for the other `useEffect` below.

MythicManiac commented on apps/sertra/components/PackageDataFetcher.tsx:
```diff
@@ -0,0 +1,65 @@
+      setMods(await (await fetchMods()).json());
+    }
+    jsonifyMods();
+  });
```
Also missing the second parameter, which in this case should simply be an empty list.

MythicManiac reviewed (changes requested):
See inline comments

Oksamies commented on apps/sertra/components/ModCard.tsx:
```diff
@@ -8,6 +9,14 @@ interface ModCard {
+  src: string;
+}
+
+const modImageLoader = ({ src }: modImageLoaderInputProps) => {
```
Good points, still getting used to finding the types 👍 Seems like the Image thing doesn't work without a width, even though its not used in the loader. The example here is a bit weird for me also (https://nextjs.org/docs/api-reference/next/image#loader), since it's just slapping the resizing at the end of the url, which I guess should be a feature on the images hosting side 🤔

Anyways I do think, the latest version (soon to be updated to this branch), is fine enough for the Alpha.

anttimaki replied:
I guess the 3rd point in my previous comment was vaguely worded (too many "it" making it unclear what I'm referring to).

What I meant was that it looks the `loader` prop isn't needed. And since your implementation just returns the `src` parameter as is, I think defining the loader func and including the prop only adds noise. So I would remove the `loader` prop if the image works as expected without it.

Oksamies replied:
Ah yes, seems like the loader can be removed as long as it has that `unoptimized` as `true`, but I'm not sure why it works with it 🤷 I tried it without the loader, before adding the unoptimized, but back then it didn't work.

Might have something to do with the fact that "optimization" is probably configured  manually in the loader, depending on the hosting parties way of delivering "optimized" images? 🤷 Anyways should be good now

anttimaki replied:
I'm not thrilled with the amount of guesswork and shrugging emoji here. Created TS-1054 for figuring this out so it's not a blocker for this PR.

Oksamies commented:
@MythicManiac @anttimaki I suggest looking at the changed files, instead of the commits, since the commits are quite a mess now. After you guys approve the changes, I'll cleanup the commit history to be something readable. Ty in advance ❤️

anttimaki reviewed (approved):
Approving the changes done based on *my comments*. The PR as a whole isn't approved yet by all reviewers.

MythicManiac commented on apps/sertra/pages/server/[id].tsx:
```diff
@@ -52,37 +57,29 @@ const ServerDetail: React.FC<{ detail_listing: ServerListingDetailData }> = ({
-              {detail_listing.mods.map((x) => (
-                <ModCard key={x} name={x} description={x} />
+              {mods_data.map((modProps) => (
+                <ModCard key={modProps.name} {...modProps} />
```
`key` must be guaranteed to be unique, so you should use the full name for it

MythicManiac commented on apps/sertra/api/PackageDataFetcher.ts:
```diff
@@ -0,0 +1,49 @@
+      all_versions.set(version_with_owner.full_name, version_with_owner);
+    });
+  });
+  const updated_listing_mods_data = [];
+  for (const mod_ref of listing.mods) {
+    const new_mod_data: ListingMod = {
+      name: "MOD DETAIL FETCH FAILED: " + mod_ref,
+      owner: null,
+      version: null,
+      icon_url: null,
+      description: null,
+    };
+    const mod = all_versions.get(mod_ref);
+    if (mod) {
+      new_mod_data.name = mod.name;
+      new_mod_data.owner = mod.owner;
+      new_mod_data.version = mod.version_number;
+      new_mod_data.icon_url = mod.icon;
+      new_mod_data.description = mod.description;
+    }
+    updated_listing_mods_data.push(new_mod_data);
```
You should be able to simplify this quite a bit, something like:

```typescript
updated_listing_mods_data = listing.mods.map((x) => all_versions.get(x) ?? {
  name: mod_ref,
  owner: null,
  version: null,
  icon_url: null,
  description: null,
});
```

MythicManiac commented on apps/sertra/api/PackageDataFetcher.ts:
```diff
@@ -0,0 +1,49 @@
+  const mods = await mod_data_fetch;
+
+  const all_versions = new Map();
+  mods.map((mod: TSListingMod) => {
+    mod.versions.map((version) => {
+      const version_with_owner = version as TSListingModVersionWithOwner;
+      version_with_owner.owner = mod.owner;
+      all_versions.set(version_with_owner.full_name, version_with_owner);
+    });
+  });
```
Here on the other hand I'd prefer that you use a for loop, as map is usually used when you need the return value (you don't here, you're just looping)

MythicManiac reviewed (approved):
Implementation-wise looks correct now, left some minor comments
